cmd/gardener-scheduler/app/gardener_scheduler.go

```diff
@@ -0,0 +1,369 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package app
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"k8s.io/client-go/discovery"
+	"os"
+	"time"
+
+	gardenv1beta1 "github.com/gardener/gardener/pkg/apis/garden/v1beta1"
+	gardencoreinformers "github.com/gardener/gardener/pkg/client/core/informers/externalversions"
+	gardeninformers "github.com/gardener/gardener/pkg/client/garden/informers/externalversions"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	gardenmetrics "github.com/gardener/gardener/pkg/controllermanager/metrics"
+	"github.com/gardener/gardener/pkg/logger"
+	"github.com/gardener/gardener/pkg/scheduler/apis/config"
+	schedulerconfigv1alpha1 "github.com/gardener/gardener/pkg/scheduler/apis/config/v1alpha1"
+	"github.com/gardener/gardener/pkg/scheduler/controller"
+	"github.com/gardener/gardener/pkg/scheduler/validation"
+	"github.com/gardener/gardener/pkg/server"
+	"github.com/gardener/gardener/pkg/server/handlers"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	diskcache "k8s.io/client-go/discovery/cached/disk"
+	"k8s.io/client-go/informers"
+	kubeinformers "k8s.io/client-go/informers"
+	k8s "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	"github.com/gardener/gardener/cmd/utils"
+)
+
+// Options has all the context and parameters needed to run a GardenerScheduler.
+type Options struct {
+	// ConfigFile is the location of the GardenerScheduler's configuration file.
+	ConfigFile string
+	config     *config.SchedulerConfiguration
+	codecs     serializer.CodecFactory
+}
+
+// AddFlags adds flags for a specific Scheduler to the specified FlagSet.
+func (o *Options) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&o.ConfigFile, "config", o.ConfigFile, "The path to the configuration file.")
+}
+
+// NewOptions returns a new Options object.
+func NewOptions() (*Options, error) {
```

Comment:
a constructor for options probably shouldn't return an `error` - Once you moved out the config decoding, you can probably drop the `error`.